dbt/adapters/base/impl.py

```diff
@@ -0,0 +1,705 @@
+from dbt.adapters.cache import RelationsCache
+
+GET_CATALOG_OPERATION_NAME = 'get_catalog_data'
+
+
+def _expect_row_value(key, row):
+    if key not in row.keys():
+        raise dbt.exceptions.InternalException(
+            'Got a row without "{}" column, columns: {}'
+            .format(key, row.keys())
+        )
+    return row[key]
+
+
+def _relations_filter_schemas(schemas):
+    def test(row):
+        referenced_schema = _expect_row_value('referenced_schema', row)
+        dependent_schema = _expect_row_value('dependent_schema', row)
+        # handle the null schema
+        if referenced_schema is not None:
+            referenced_schema = referenced_schema.lower()
+        if dependent_schema is not None:
+            dependent_schema = dependent_schema.lower()
+        return referenced_schema in schemas or dependent_schema in schemas
+    return test
+
+
+def _catalog_filter_schemas(manifest):
+    """Return a function that takes a row and decides if the row should be
+    included in the catalog output.
+    """
+    schemas = frozenset(s.lower() for s in manifest.get_used_schemas())
+
+    def test(row):
+        table_schema = _expect_row_value('table_schema', row)
+        # the schema may be present but None, which is not an error and should
+        # be filtered out
+        if table_schema is None:
+            return False
+        return table_schema.lower() in schemas
+    return test
+
+
+@six.add_metaclass(AdapterMeta)
+class BaseAdapter(object):
+    """The BaseAdapter provides an abstract base class for adapters.
+
+    Adapters must implement the following methods. Some of these methods can be
+    safely overridden as a noop, where it makes sense (transactions on
+    databases that don't support them, for instance). Those methods are marked
+    with a (passable) in their docstrings. Check docstrings for type
+    information, etc.
+
+    Methods:
+        - exception_handler
+        - date_function
+        - get_existing_schemas
+        - drop_relation
+        - truncate_relation
+        - rename_relation
+        - get_columns_in_relation
+        - expand_column_types
+        - list_relations_without_caching
+        - is_cancelable
+        - create_schema
+        - drop_schema
+        - quote
+        - convert_text_type
+        - convert_number_type
+        - convert_boolean_type
+        - convert_datetime_type
+        - convert_date_type
+        - convert_time_type
+    """
+    requires = {}
+
+    Relation = BaseRelation
+    Column = Column
+    # This should be an implementation of BaseConnectionManager
+    ConnectionManager = None
+
+    def __init__(self, config):
+        self.config = config
+        self.cache = RelationsCache()
+        self.connections = self.ConnectionManager(config)
+
+    ###
+    # Methods that pass through to the connection manager
+    ###
+    # TODO: I would like a better way to do this. Should adapters be subclasses
```

Comment:
Can you say more about what a better version of this code looks like?

Reply:
I should've removed this comment. I think it makes the class hierarchy very hard, but it does make using the classes nicer. Basically, the per-database connection managers would derive from the base connection manager and the base adapter would derive from the base connection manager as well. All adapter implementations would then derive from both the base adapter and their database-specific connection manager. Inheritance would be a mess but python's method resolution model would take care of it nicely. And we could remove the methods below this comment.

It's probably better to just preserve this small bit of ugliness.

Reply:
```
BaseConnectionManager
               / \
              /   \
             /     \
   BaseAdapter    SQLConnectionManager
            \          /    \
             \        /      \
              \      /        \
             SQLAdapter    PostgresConnectionManager
                \                |
                 \   RedshiftConnectionManager
                  \    /                 
                 RedshiftAdapter
```

An ascii diagram of what (some of) the inheritance model would look like (this was so hard to make)